src/v/kafka/server/group_manager.cc

```diff
@@ -614,6 +628,104 @@ group_manager::txn_offset_commit(txn_offset_commit_request&& r) {
+    return reply;
+}
+
+ss::future<cluster::begin_group_tx_reply>
+group_manager::begin_tx(cluster::begin_group_tx_request&& r) {
+    if (!_catchup_lock.take_reader_lock()) {
+        vlog(klog.warn, "can't process a tx: coordinator_load_in_progress");
+        co_return make_begin_tx_reply(
+          cluster::tx_errc::coordinator_load_in_progress);
+    }
+
+    try {
+        auto error = validate_group_status(
+          r.ntp, r.group_id, offset_commit_api::key);
+        if (error != error_code::none) {
+            _catchup_lock.release_reader_lock();
+            if (error == error_code::not_coordinator) {
+                co_return make_begin_tx_reply(
+                  cluster::tx_errc::not_coordinator);
+            } else {
+                co_return make_begin_tx_reply(cluster::tx_errc::timeout);
+            }
+        }
+
+        auto partition = _partitions.find(r.ntp)->second;
+
+        if (partition->partition->term() != partition->term) {
+            _catchup_lock.release_reader_lock();
+            co_return make_begin_tx_reply(cluster::tx_errc::not_coordinator);
+        }
+
+        auto fence_it = partition->fence_pid_epoch.find(r.pid.get_id());
+        if (
+          fence_it == partition->fence_pid_epoch.end()
+          || r.pid.get_epoch() > fence_it->second) {
+            // TODO: https://app.clubhouse.io/vectorized/story/2200
+            // include producer_id into key to make it unique-ish
+            // to prevent being GCed by the compaction
+            auto batch = cluster::make_fence_batch(
+              fence_control_record_version, r.pid);
+            auto reader = model::make_memory_record_batch_reader(
+              std::move(batch));
+            auto e = co_await partition->partition->replicate(
+              partition->term,
+              std::move(reader),
+              raft::replicate_options(raft::consistency_level::quorum_ack));
+
+            if (!e) {
+                vlog(
+                  klog.error,
+                  "Error \"{}\" on replicating pid:{} fencing batch",
+                  e.error(),
+                  r.pid);
+                _catchup_lock.release_reader_lock();
+                co_return make_begin_tx_reply(cluster::tx_errc::timeout);
+            }
+
+            fence_it = partition->fence_pid_epoch.find(r.pid.get_id());
+        }
+
+        if (fence_it == partition->fence_pid_epoch.end()) {
+            partition->fence_pid_epoch.emplace(
+              r.pid.get_id(), r.pid.get_epoch());
+        } else if (r.pid.get_epoch() >= fence_it->second) {
+            fence_it->second = r.pid.get_epoch();
+        } else {
+            vlog(
+              klog.error,
+              "pid {} fenced out by epoch {}",
+              r.pid,
+              fence_it->second);
+            _catchup_lock.release_reader_lock();
+            co_return make_begin_tx_reply(cluster::tx_errc::fenced);
+        }
+
+        auto group = get_group(r.group_id);
```

Comment:
is this line actually executed? the `else` above seems to always return tx_errc::fenced no?

Reply:
yep, the `else` always returns but the other branched don't